### ISSUE TITLE
Consider deprecating `@layer.mouse_drag_callbacks.append` syntax

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#2267]
@sofroniewn got me thinking about mouse & key callbacks recently, both of which pose small (but solvable) problems for trimming down our models and Viewer/Layer objects to their "essential data".  Currently, we store callbacks on the classes and instances themselves, in special dicts and lists for the purposes.

The primary advantage I can see is the nifty decorator that we have in our examples:

```python
my_layer = viewer.add_points()

@my_layer.mouse_drag_callbacks.append
def my_callback(): ...
```

These attributes aren't exactly fundamental data about the layer or viewer though, and their mixin classes (the classes that actually add the `mouse_drag_callbacks`, etc... attributes to the subclass) make it a bit difficult to use them in pydantic-style classes.  Additionally, it scatters our key/mouse-binding information across a lot of different objects, making it a bit harder to ask "what's bound to the `C` key?"

One option would be to make those mixins purely _behavioral_ (methods only), where the `bind_mouse_callback` method would store the actual event/callback binding elsewhere in some more global mapping.  That solves a handful of problems, but it means we'd need to change the example above to something like:

```python
my_layer = viewer.add_points()

@my_layer.bind_mouse_drag_callback  # no "append" here
def my_callback(): ...
```

thoughts?

**Comment by sofroniewn** [comment]
> One option would be to make those mixins purely behavioral (methods only), where the bind_mouse_callback method would store the actual event/callback binding elsewhere in some more global mapping. 

Yes, I prefer using methods here and think it will give us greater control. It will also bring things closer in line with our `@bind_key` approach.

I wonder when we do this if it isn't also an opportunity to look at some of the proposals in #1048 again too, for example we could do something like `@bind_mouse('press')` too.

Ideally we could do this in a way with deprecations. If so then I say full steam ahead. I'm also thinking about how to let plugins provide interactive callbacks, but will make a new issue for that.
